### PR TITLE
Add Numba-dppy Demo Notebook for Numba 0.54 release

### PR DESCRIPTION
**Pull request by reazulhoque**
As Title.

### CODE REVIEW

**Comment by sklam** [comment]
Getting some trouble with conda install. What i am doing:

```bash
$ conda create -n numba-dppy-env numba-dppy dpnp numba::numba=0.54.0 -c intel
Collecting package metadata (current_repodata.json): done
Solving environment: failed with repodata from current_repodata.json, will retry with next repodata source.
Collecting package metadata (repodata.json): done
Solving environment: \
Found conflicts! Looking for incompatible packages.
This can take several minutes.  Press CTRL-C to abort.
failed

UnsatisfiableError: The following specifications were found to be incompatible with each other:

Output in format: Requested package -> Available versions

Package mkl conflicts for:
dpnp -> mkl[version='>=2021.1.1|>=2021.1b10']
dpnp -> mkl-dpcpp[version='>=2021.1.1'] -> mkl[version='2021.1.1|2021.2.0|2021.3.0|>=2021.3.0,<2022.0a0|>=2021.2.0,<2022.0a0|>=2019.4,<2021.0a0|>=2019.3,<2021.0a0|>=2019.1,<2021.0a0|>=2018.0.3,<2019.0a0|>=2019.4,<2020.0a0',build='intel_520|intel_296|intel_52']

Package libgcc-ng conflicts for:
dpnp -> libgcc-ng[version='>=9.3.0']
dpnp -> numpy[version='>=1.15'] -> libgcc-ng[version='>=7.2.0|>=7.3.0|>=7.5.0']

Package _libgcc_mutex conflicts for:
dpnp -> libgcc-ng[version='>=9.3.0'] -> _libgcc_mutex[version='*|0.1',build=main]
numba-dppy -> libgcc-ng[version='>=9.3.0'] -> _libgcc_mutex[version='*|0.1',build=main]
numba::numba=0.54.0 -> libgcc-ng[version='>=7.5.0'] -> _libgcc_mutex[version='*|0.1',build=main]

Package numba conflicts for:
numba::numba=0.54.0
numba-dppy -> numba[version='0.52.*|0.53.*']

Package numpy conflicts for:
dpnp -> numpy[version='>=1.15|>=1.20,<1.21.0a0|>=1.19,<1.20.0a0']
dpnp -> dpctl[version='>=0.5.0a0'] -> numpy[version='>=1.17']

Package python conflicts for:
dpnp -> numpy[version='>=1.15'] -> python[version='>=2.7,<2.8.0a0|>=3.6,<3.7.0a0|>=3.9,<3.10.0a0|>=3.5,<3.6.0a0']
dpnp -> python[version='>=3.7,<3.8.0a0|>=3.8,<3.9.0a0']

Package libstdcxx-ng conflicts for:
dpnp -> libstdcxx-ng[version='>=9.3.0']
dpnp -> python[version='>=3.7,<3.8.0a0'] -> libstdcxx-ng[version='>=7.2.0|>=7.3.0|>=7.5.0']

Package numpy-base conflicts for:
numba::numba=0.54.0 -> numpy[version='>=1.17,<1.21'] -> numpy-base[version='1.17.0|1.17.0|1.17.0|1.17.0|1.17.0|1.17.4|1.17.5|1.18.5|1.18.5|1.19.2|1.20.1|1.20.3|1.20.3|1.20.3|1.20.3|1.20.3|1.20.3|1.20.3|1.20.2|1.20.2|1.20.2|1.20.2|1.20.2|1.20.2|1.20.1|1.20.1|1.20.1|1.20.1|1.20.1|1.20.1|1.19.2|1.19.2|1.19.2|1.19.2|1.19.2|1.19.2|1.19.2|1.19.2|1.19.1|1.19.1|1.19.1|1.19.1|1.19.1|1.19.1|1.18.5.*|1.18.1.*|1.17.4.*|1.17.3.*|1.17.2.*',build='py36h2f8d375_0|py37hde5b4d6_0|py38h75fe3a5_0|py38hfa32c7d_0|py36hfa32c7d_0|py37hfa32c7d_0|py38h75fe3a5_0|py37h75fe3a5_0|py36hfa32c7d_0|py39h0f7b65f_0|py37h7d8b39e_0|py39h34387ca_0|py38h7d8b39e_0|py39h7d8b39e_0|py38hfae3a4d_0|py39he2ba247_0|py37hfae3a4d_0|py39h74d4b33_0|py37h39b7dee_0|py37hf11705a_19|py37hf45626f_1|py38hf45626f_1|py37hffe281f_0|py37h591eb60_0|py37_5|py37_1|py37_4|py36_13|py39h39b7dee_0|py38h39b7dee_0|py38h74d4b33_0|py37h74d4b33_0|py39hfae3a4d_0|py37he2ba247_0|py38he2ba247_0|py38h34387ca_0|py37h34387ca_0|py39h2ae0177_0|py38hfa32c7d_0|py36h75fe3a5_0|py37h75fe3a5_0|py36h75fe3a5_0|py37hfa32c7d_0|py36hde5b4d6_0|py37h2f8d375_0']
dpnp -> numpy[version='>=1.15'] -> numpy-base[version='1.15.0|1.15.0|1.15.0|1.15.0|1.15.0|1.15.0|1.15.0|1.15.0|1.15.1|1.15.1|1.15.1|1.15.1|1.15.1|1.15.1|1.15.1|1.15.1|1.15.1|1.15.1|1.15.1|1.15.1|1.15.2|1.15.2|1.15.2|1.15.2|1.15.2|1.15.2|1.15.2|1.15.2|1.15.2|1.15.2|1.15.2|1.15.2|1.15.2|1.15.2|1.15.2|1.15.2|1.15.3|1.15.3|1.15.4|1.15.4|1.16.1|1.16.1|1.16.2|1.16.2|1.16.4|1.17.0|1.17.4|1.17.5|1.18.5|1.18.5|1.19.2|1.20.1|1.20.3|1.20.3|1.20.3|1.20.3|1.20.3|1.20.3|1.20.3|1.20.2|1.20.2|1.20.2|1.20.2|1.20.2|1.20.2|1.20.1|1.20.1|1.20.1|1.20.1|1.20.1|1.20.1|1.19.2|1.19.2|1.19.2|1.19.2|1.19.2|1.19.2|1.19.2|1.19.2|1.19.1|1.19.1|1.19.1|1.19.1|1.19.1|1.19.1|1.18.5.*|1.18.1.*|1.17.4.*|1.17.3.*|1.17.2.*|1.16.6|1.16.6|1.16.6|1.16.6|1.16.6|1.16.6|1.16.6|1.16.6|1.16.6|1.16.6|1.16.6|1.16.6|1.16.6|1.16.6|1.16.6|1.16.6|1.16.6|1.16.6|1.16.5|1.16.5|1.16.5|1.16.5|1.16.5|1.16.5|1.16.4|1.16.4|1.16.4|1.16.4|1.16.4|1.16.4|1.16.3|1.16.3|1.16.3|1.16.3|1.16.3|1.16.3|1.16.2|1.16.2|1.16.2|1.16.2|1.16.2|1.16.2|1.16.1|1.16.1|1.16.1|1.16.1|1.16.1|1.16.1|1.16.1|1.16.1|1.16.1|1.16.1|1.16.1|1.16.1|1.16.0|1.16.0|1.16.0|1.16.0|1.16.0|1.16.0|1.16.0|1.16.0|1.16.0|1.16.0|1.16.0|1.16.0|1.15.4|1.15.4|1.15.4|1.15.4|1.15.4|1.15.4|1.15.4|1.15.4|1.15.4|1.15.3|1.15.3|1.15.3|1.15.3|1.15.3|1.15.3|1.17.0|1.17.0|1.17.0|1.17.0',build='py36h2f8d375_0|py37h2f8d375_0|py37hde5b4d6_0|py27h7cdd4dd_0|py37h7cdd4dd_0|py35h7cdd4dd_0|py27h3dfced4_0|py35h3dfced4_0|py35h74e8950_0|py27h74e8950_0|py36h74e8950_0|py37h74e8950_0|py36h81de0dd_0|py37h81de0dd_0|py35h81de0dd_0|py35h2f8d375_0|py27h2f8d375_0|py36h2f8d375_0|py27h2f8d375_0|py27h81de0dd_0|py35h81de0dd_0|py35h2f8d375_0|py36h2f8d375_0|py36h81de0dd_0|py36h2f8d375_1|py27h2f8d375_1|py27h81de0dd_1|py27h2f8d375_0|py36h2f8d375_0|py27h81de0dd_0|py37h81de0dd_0|py37h2f8d375_0|py36h2f8d375_0|py27h2f8d375_0|py36h81de0dd_0|py27h81de0dd_0|py37hde5b4d6_0|py27h2f8d375_0|py37h2f8d375_0|py37hde5b4d6_0|py36h2f8d375_0|py27hde5b4d6_0|py27h2f8d375_1|py27hde5b4d6_1|py36hde5b4d6_1|py37h2f8d375_0|py27h2f8d375_0|py36h2f8d375_0|py37hde5b4d6_0|py36hde5b4d6_1|py27hde5b4d6_1|py36h2f8d375_0|py27h2f8d375_0|py37hde5b4d6_0|py27h2f8d375_0|py36h2f8d375_0|py27h2f8d375_0|py36h2f8d375_0|py37h2f8d375_0|py37hde5b4d6_0|py27hde5b4d6_0|py36h2f8d375_0|py27h2f8d375_0|py36h2f8d375_0|py38hde5b4d6_0|py27h2f8d375_0|py39h76555f2_1|py39h41b4c56_3|py36hdc34a94_3|py39hdc34a94_3|py38h75fe3a5_0|py38hfa32c7d_0|py36hfa32c7d_0|py37h75fe3a5_0|py37hfa32c7d_0|py38h75fe3a5_0|py36hfa32c7d_0|py38hfa32c7d_0|py39h34387ca_0|py38h34387ca_0|py38h7d8b39e_0|py39h7d8b39e_0|py37he2ba247_0|py37h74d4b33_0|py39h74d4b33_0|py37h39b7dee_0|py36_0|py36_0|py36_0|py27_25|py37hf11705a_19|py37hf45626f_1|py38hf45626f_1|py37hffe281f_0|py37h591eb60_0|py37_5|py37_1|py37_4|py36_13|py27_0|py36_3|py27_3|py36_2|py27_2|py27_0|py27_0|py39h39b7dee_0|py38h39b7dee_0|py38h74d4b33_0|py39hfae3a4d_0|py37hfae3a4d_0|py39he2ba247_0|py38hfae3a4d_0|py38he2ba247_0|py37h7d8b39e_0|py37h34387ca_0|py39h0f7b65f_0|py39h2ae0177_0|py36h75fe3a5_0|py37h75fe3a5_0|py36h75fe3a5_0|py37hfa32c7d_0|py37hdc34a94_3|py38hdc34a94_3|py36h41b4c56_3|py37h41b4c56_3|py38h41b4c56_3|py39hfb011de_1|py27hde5b4d6_0|py38h2f8d375_0|py36hde5b4d6_0|py37hde5b4d6_0|py37h2f8d375_0|py36hde5b4d6_0|py37hde5b4d6_0|py27hde5b4d6_0|py37h2f8d375_0|py36hde5b4d6_0|py36hde5b4d6_0|py27hde5b4d6_0|py37hde5b4d6_0|py37h2f8d375_0|py36hde5b4d6_0|py27hde5b4d6_0|py37h2f8d375_0|py37hde5b4d6_1|py36h2f8d375_1|py37h2f8d375_1|py27h2f8d375_1|py27hde5b4d6_0|py36hde5b4d6_0|py37hde5b4d6_1|py36h2f8d375_1|py37h2f8d375_1|py36hde5b4d6_0|py36hde5b4d6_0|py27hde5b4d6_0|py37h81de0dd_0|py36h81de0dd_0|py37h2f8d375_0|py37h2f8d375_1|py37h81de0dd_1|py36h81de0dd_1|py37h81de0dd_0|py37h2f8d375_0|py37h2f8d375_0|py27h81de0dd_0|py37h3dfced4_0|py36h3dfced4_0|py36h7cdd4dd_0|py36hde5b4d6_0']

Package dpnp conflicts for:
numba-dppy -> dpnp[version='>=0.5.1|>=0.6*,<0.7*']
dpnpThe following specifications were found to be incompatible with your system:

  - feature:/linux-64::__glibc==2.28=0
  - feature:|@/linux-64::__glibc==2.28=0
  - dpnp -> libgcc-ng[version='>=9.3.0'] -> __glibc[version='>=2.17']
  - numba-dppy -> libgcc-ng[version='>=9.3.0'] -> __glibc[version='>=2.17']

Your installed version is: 2.28
```

**Comment by sklam** [comment]
@reazulhoque, looks like we need `numba-dppy=0.15` but it's not available in https://anaconda.org/intel/numba-dppy/files yet

**Comment by PokhodenkoSA** [comment]
Try `-c dppy/label/dev`. numba-dppy is not in `-c intel` yet.

**Comment by sklam** [comment]
@PokhodenkoSA 

There's a glibc ABI problem:

```bash
$ conda create -n dppy -c dppy/label/dev numba-dppy
Collecting package metadata (current_repodata.json): done
Solving environment: failed with repodata from current_repodata.json, will retry with next repodata source.
Collecting package metadata (repodata.json): done
Solving environment: -
Found conflicts! Looking for incompatible packages.
This can take several minutes.  Press CTRL-C to abort.
failed

UnsatisfiableError: The following specifications were found to be incompatible with your system:

  - feature:/linux-64::__glibc==2.28=0
  - numba-dppy -> libgcc-ng[version='>=9.3.0'] -> __glibc[version='>=2.17']

Your installed version is: 2.28
```

**Comment by PokhodenkoSA** [comment]
@sklam 
Please, try this: `-c defaults -c numba -c intel -c numba/label/dev -c dppy/label/dev --override-channels`.

**Comment by sklam** [comment]
@PokhodenkoSA  I did

```
$ conda create -n dppy numba-dppy==0.15.0 -c defaults -c numba -c intel -c numba/label/dev -c dppy/label/dev --override-channels
```

but same problem

**Comment by sklam** [comment]
At one point, we have a similar problem with numba. we need to add https://github.com/numba/numba/blob/6c01bbcb36a75f6c6bce666601f3c879ba6e5136/buildscripts/condarecipe.local/conda_build_config.yaml to force the use of gcc 7 series.

**Comment by PokhodenkoSA** [comment]
Fast and simple and stable way it:
```
git clone https://github.com/IntelPython/numba-dppy.git
cd numba-dppy
conda env create -f environment.yml
conda activate dev
python setup.py develop
```
Sorry for asking to build from sources. This is something that is guaranteed to work and fast.

**Comment by PokhodenkoSA** [comment]
It is while I try to find a solution for he package.

File: notebooks/Numba_054_DPPY_Release_Demo.ipynb [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+    "\n",
+    "N = 64 * 32\n",
+    "a = np.ones(N, dtype=np.float32)\n",
+    "b = np.ones(N, dtype=np.float32)\n",
+    "\n",
+    "# Use the environment variable SYCL_DEVICE_FILTER to change the default device.\n",
+    "# See https://github.com/intel/llvm/blob/sycl/sycl/doc/EnvironmentVariables.md#sycl_device_filter.\n",
+    "device = dpctl.select_default_device()\n",
+    "print(\"Using device ...\")\n",
+    "device.print_device_info()\n",
```
Comment (sklam) [inline notebooks/Numba_054_DPPY_Release_Demo.ipynb:70]: This is print is not showing up in the notebook

File: notebooks/Numba_054_DPPY_Release_Demo.ipynb [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+    "    R[i] = A[i] + A[i + stride]\n",
+    "    # store the sum to be used in nex iteration\n",
+    "    A[i] = R[i]\n",
+    "\n",
+    "\n",
+    "def sum_reduce(A):\n",
+    "    \"\"\"Size of A should be power of two.\"\"\"\n",
+    "    total = len(A)\n",
+    "    # max size will require half the size of A to store sum\n",
+    "    R = np.array(np.random.random(math.ceil(total / 2)), dtype=A.dtype)\n",
```
Comment (sklam) [inline notebooks/Numba_054_DPPY_Release_Demo.ipynb:279]: `math` is never imported

**Review by sklam — COMMENTED** [review]
I tested it with manually installing `numba-dppy` from source and running in a docker container.

File: notebooks/Numba_054_DPPY_Release_Demo.ipynb (hunk 1 above)
Comment (sklam) [reply notebooks/Numba_054_DPPY_Release_Demo.ipynb:70]: maybe changing to `print(device)`, which is printing:

```
<dpctl.SyclDevice [backend_type.opencl, device_type.cpu,  Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz] at 0x7fb63152d370>
```
for me.

File: notebooks/Numba_054_DPPY_Release_Demo.ipynb [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+    "    NUMBA_DPPY_ACTIVATE_ATOMCIS_FP_NATIVE=1\n",
+    "    NUMBA_DPPY_LLVM_SPIRV_ROOT=/path/to/dpcpp/provided/llvm_spirv\n",
+    "\n",
+    "    To run this example:\n",
+    "    NUMBA_DPPY_ACTIVATE_ATOMCIS_FP_NATIVE=1 \n",
+    "    NUMBA_DPPY_LLVM_SPIRV_ROOT=/path/to/dpcpp/provided/llvm_spirv python atomic_op.py\n",
+    "\n",
+    "    Without these two environment variables Numba_dppy will use other\n",
+    "    implementation for floating point atomics.\n",
+    "    \"\"\"\n",
```
Comment (sklam) [inline notebooks/Numba_054_DPPY_Release_Demo.ipynb:172]: The docstring seems to imply that the atomic only works on GPU devices, but it appears to be working on the CPU as well. Also, the description needs to be adapted for the notebook; e.g. the file `atomic_op.py` do not exist.

**Comment by sklam** [comment]
Note: 

use the following command to install `numba-dppy`
`conda create -n dppy numba-dppy=0.16 -c defaults -c numba -c intel -c numba/label/dev -c dppy/label/dev -c conda-forge --override-channels`

**Comment by sklam** [comment]
@reazulhoque, I've reviewed it and i have a few suggestions

**Comment by reazulhoque** [comment]
Thank you @sklam! I will try to address the suggestions asap.

**Pull request merged**